### ISSUE TITLE
ml: train and validate a physics-residual pedestrian goal-belief model

### ISSUE THREAD

**Issue by ll7** [ll7/robot_sf_ll7#8098]
## Current issue-planning review — 2026-09-15

**Audit disposition:** `blocked`. The current audit classifies the issue as blocked; its prerequisite, evidence, or triage gate is incomplete.

**First action:** Train, calibrate, and validate a compact physics-residual learned model that improves the best confirmed rule-based pedestrian goal-force/route belief while preserving interpretability, actor-only inputs, causal histories, deterministic checkpoint provenance, and safe fallback behavior.

### Current bounded plan

1. Inspect the named owners and dependencies before editing; keep the live issue scope and fail-closed evidence boundary intact.
2. Run the named focused validation, record exact output and remaining blockers, and keep fallback, degraded, unavailable, and inconclusive outcomes explicit.

### Files and owners to inspect

- robot_sf/prediction/goal_belief_dataset.py
- scripts/data/build_goal_belief_dataset.py
- robot_sf/prediction/learned_goal_belief.py
- robot_sf/prediction/learned_goal_belief_trainer.py
- scripts/training/train_goal_belief.py
- configs/training/goal_belief_gru_v1.yaml
- tests/prediction/test_goal_belief_dataset.py
- tests/prediction/test_learned_goal_belief.py
- tests/prediction/test_learned_goal_belief_trainer.py
- tests/prediction/test_goal_belief_checkpoint_contract.py
- tests/prediction/test_goal_belief_calibration.py
- tests/prediction/test_goal_belief_online_adapter.py
- robot_sf/prediction
- scripts/data
- scripts/training
- tests/prediction
- scripts/benchmark/run_goal_force_inference.py
- scripts/dev/pr_ready_check.sh

### Acceptance and validation

- Demonstrate the stated outcome from the live issue with a focused, reproducible check.
- Keep any missing prerequisite or inconclusive result explicit rather than treating it as success.

### Current dependencies and issue-text corrections

- No supplied individual draft existed for this issue in the handoff package; this is a conservative plan derived from the refreshed live body, and the historical issue text is preserved below.
- No explicit issue dependency was found in the live body; establish the canonical owner and prerequisites before implementation.
- Keep current out-of-scope, ownership, provenance, rights, compute, and claim boundaries intact. Do not close the issue merely because this plan was posted.

### Lifecycle and claim boundary

- Current state labels: `none`.
- The issue remains open. No closure, compute submission, release/publication, evidence deletion, or scientific-claim promotion is authorized by this review.
- This review does not request a lifecycle transition; existing assignees, ownership, valid priority labels, and unrelated labels are preserved.
- Proposed non-lifecycle label additions applied with this review: `none`.

**Evidence note:** Full live issue body and complete available comment thread were reviewed during the 2026-09-15 refresh. This is planning metadata, not implementation or benchmark evidence; historical measurements remain historical.

## Goal

Train, calibrate, and validate a compact physics-residual learned model that improves the best confirmed rule-based pedestrian goal-force/route belief while preserving interpretability, actor-only inputs, causal histories, deterministic checkpoint provenance, and safe fallback behavior.

The first learned backend should be a small GRU or equivalently bounded recurrent/temporal model. It should learn residual model error and transition cues, not replace the known Social Force decomposition with an opaque end-to-end trajectory model.

## Dependencies

Blocked by:

- `contract: define GoalBeliefV1, oracle trace, timing, and leakage boundary`
- `sim: instrument side-effect-free pedestrian force and goal-transition oracle traces`
- `observation: add stable pedestrian tracking and canonical frame normalization`
- `prediction: implement multi-horizon inverse pedestrian goal-force estimator`
- `prediction: generate non-oracle goal candidates and hierarchical waypoint/route belief`
- `prediction: detect waypoint progress, arrival, redirects, and reset goal beliefs`
- `benchmark: build calibrated goal-force inference evaluation and robustness harness`
- `integration: expose observation-only GoalBelief through ScenarioBelief and forecasts`

A rule baseline and benchmark confirmation configuration must be frozen before model selection or final test evaluation.

Parent: `research: observation-only pedestrian goal-force and hierarchical intention inference`

## Research hypothesis

A learned residual model may improve probabilistic goal-force, waypoint/route, and change predictions by modelling:

- parameter mismatch in preferred speed, relaxation time, and force factors;
- unobserved or partially observed interactions;
- tracking and differentiation noise;
- transition patterns around waypoint arrival, stopping, restart, and redirects;
- data-supported candidate priors;
- nonlinear relationships between force innovations and route ambiguity.

It should not be tasked with relearning force components already known and reconstructable from actor-visible inputs.

### 1. Frozen rule baseline registration

Before training, register:

- exact best rule-based config hash;
- source commit;
- observation/candidate/trace schema versions;
- development and confirmation partitions;
- baseline metrics and runtime;
- known degraded/unavailable cases;
- candidate recall ceiling.

The learned model's test results must be compared against this exact baseline, not a retrospectively selected weaker comparator.

### 2. Dataset builder

Create a dataset from actor-visible feature histories joined offline to evaluator-only labels. Suggested paths:

```text
robot_sf/prediction/goal_belief_dataset.py
scripts/data/build_goal_belief_dataset.py
```

Each example must contain two physically separated records:

#### Actor features

- causal tracked positions and velocities;
- actual timestamps and history mask;
- acceleration mean/covariance;
- reconstructed observable force components and status;
- residual rule-based goal-force estimate/covariance;
- robot-relative state/action available under the observation contract;
- local neighbor features or graph with visibility masks;
- candidate geometry, route signatures, priors, and feasibility;
- current rule posterior and change-detector evidence;
- tracking confidence, occlusion, saturation, and model-mismatch flags;
- configured public parameters and their uncertainty;
- map/scenario identifiers used only for partitioning/provenance, not as shortcut labels unless explicitly justified.

#### Oracle labels

- true desired-force vector at force-evaluation time;
- true active waypoint/candidate match or absent-candidate label;
- true route/final destination where defined;
- event/change class and transition step;
- speed-cap/goal-threshold truth;
- next-step kinematics for consistency loss.

The serialized actor tensor must not contain oracle coordinates/IDs in unused columns or metadata passed to the model.

### 3. Dataset manifest and provenance

Write a versioned manifest containing:

- source commit and dirty-tree status;
- generator config/hash;
- schema versions;
- scenario/map/seed identities;
- partition assignment;
- feature normalization statistics computed from training only;
- class counts and missingness;
- oracle label definitions/timing;
- checksums for all shards;
- license/provenance for any external real-trajectory data;
- claim boundary.

The loader must fail closed on checksum, schema, feature-order, normalization, or partition drift.

### 4. Partition strategy

No adjacent frames or windows from one episode may cross partitions. Split hierarchically by:

- map;
- scenario family;
- route/homotopy family;
- seed/episode;
- goal-event pattern.

Required held-out groups include:

- unseen route branches/homotopies;
- unseen goal locations or destination zones where possible;
- absent-candidate cases;
- waypoint transitions;
- redirects;
- stop/restart behavior;
- occlusion/reacquisition;
- density shifts;
- force-parameter and preferred-speed shifts;
- pedestrian-model variants.

Use three logical stages:

```text
training
validation/model selection
frozen confirmation test
```

The final test partition cannot be used to choose architecture, loss weights, thresholds, temperature, or early stopping.

### 5. Model architecture

Implement a bounded recurrent model, initially:

```text
per-step actor feature encoder
      -> GRU over causal history
      -> optional candidate attention/scoring head
      -> multi-task residual heads
```

Required outputs:

```text
goal_force_residual_mean_xy
goal_force_covariance or mixture parameters
direction distribution parameters
candidate/route logits including unknown
arrival probability
change-class probabilities
change probability
optional preferred-speed/tau posterior
```

The final predicted force should be:

```text
f_goal_pred = f_goal_rule + delta_f_learned
```

or an equivalently explicit residual composition. Record both base and residual outputs.

Candidate count is variable. Use masked set/attention scoring, a graph representation, or a bounded top-K with explicit truncation provenance. Do not encode candidate identity as an arbitrary positional class that changes across maps.

### 6. Random history truncation

Train one model to support effective histories:

```text
H in {1,2,3,5,8,16}
```

Randomly truncate/mask training histories and provide actual `dt`/history masks. The H1 case must not receive acceleration or transition features that require an earlier observation. Evaluation must still report every H separately plus fused/automatic history behavior.

### 7. Loss function

Use a documented multi-task objective, for example:

```text
L = lambda_force * force_NLL
  + lambda_direction * circular_direction_loss
  + lambda_candidate * candidate_NLL
  + lambda_route * route_NLL
  + lambda_unknown * unknown_binary_or_multiclass_loss
  + lambda_change * change_class_loss
  + lambda_arrival * arrival_loss
  + lambda_dynamics * next_velocity_position_consistency
  + lambda_calibration * calibration_regularizer
  + lambda_residual * residual_magnitude_regularizer
```

Requirements:

- loss weights frozen from validation, never final test;
- class imbalance handled through documented weighting/sampling;
- censored force magnitude masked or treated with an appropriate censored likelihood;
- unavailable labels masked with denominator counts;
- covariance parameterization guarantees positive definiteness;
- force/direction target timing matches the active wrapper's post-behavior, pre-integration force state;
- no loss uses future observations beyond the declared prediction target.

### 8. Training implementation

Suggested paths:

```text
robot_sf/prediction/learned_goal_belief.py
robot_sf/prediction/learned_goal_belief_trainer.py
scripts/training/train_goal_belief.py
configs/training/goal_belief_gru_v1.yaml
```

Training requirements:

- deterministic seeds for Python/NumPy/Torch where supported;
- device and library versions recorded;
- CPU smoke mode plus GPU-capable full training;
- checkpoint every configured interval with best-validation selection rule;
- resumable training with optimizer/RNG state;
- no silent fallback to untrained or constant-output model in evidence runs;
- train/validation curves and final selected epoch recorded;
- failure on non-finite loss/gradients;
- bounded gradient clipping with config;
- model parameter count and inference latency reported.

### 9. Checkpoint and model registry contract

Register the selected model through existing model-registry conventions. Manifest must include:

- model ID/version;
- checkpoint SHA-256;
- architecture/config hash;
- feature schema/order;
- normalization hash;
- training dataset manifest hash;
- baseline config hash;
- source commit;
- training seed/device;
- evidence tier;
- compatibility versions;
- known limitations and fallback.

Inference must reject an incompatible checkpoint rather than attempting partial load or silently changing features.

### 10. Calibration

Calibrate on validation only using an explicit method, such as:

- temperature scaling for candidate/change logits;
- covariance scale calibration;
- isotonic or vector scaling only if justified and held-out sample size supports it.

Report pre/post calibration on validation and one untouched final test result. Calibration parameters are part of the checkpoint artifact/hash.

### 11. Baselines and ablations

Compare on identical held-out examples:

```text
H1 heading baseline
H2 inverse-force baseline
best confirmed temporal/hierarchical rule model
learned model without force decomposition
learned model without candidate geometry
learned model without change head
learned model without uncertainty head or with fixed variance
full physics-residual model
```

The no-force-decomposition ablation determines whether the physics structure contributes beyond model capacity. Do not select different learned architectures per subgroup after seeing test results.

### 12. Metrics and gates

Use all benchmark metrics from the predecessor. Headline learned-model gates must be specified before final test, including:

- proper probabilistic score improvement over the best rule baseline;
- no material degradation in candidate recall-conditioned and unconditional metrics;
- calibrated 50/90/95% coverage within predefined tolerance;
- no worse absent-candidate/unknown detection beyond tolerance;
- no worse false-reset rate or unacceptable delay;
- inference runtime and memory within the declared planner budget;
- fallback/degraded rate not increased by incompatibility or missing features.

A point-error improvement with worse NLL/calibration does not satisfy promotion.

### 13. Robustness and transfer

Run the frozen model on held-out robustness sweeps:

- noise/occlusion;
- density;
- parameter shifts;
- model variants;
- unseen maps/routes;
- candidate incompleteness;
- time-step variation within supported contract.

Report non-transfer and negative results. Do not fine-tune on the robustness test set before reporting the initial result.

### 14. Online adaptation prototype

Only after frozen offline evaluation, add an optional shadow-only adapter for safe online calibration/residual updates. It may use delayed observed one-step kinematic errors to update a small bounded layer or calibration parameters.

It may not:

- train the destination head on its own argmax;
- consume oracle goals at runtime;
- update the frozen base weights during benchmark comparison;
- persist an adapted model without a new manifest/hash;
- affect planner actions until separately promoted.

Required safeguards:

- bounded replay buffer;
- update-rate and norm limits;
- shadow pre/post metrics;
- rollback on calibration or validation regression;
- drift diagnostics;
- explicit disabled default.

### Dataset leakage

- actor feature tensors contain no oracle goal/route/ID columns;
- randomizing oracle labels after actor feature generation leaves features byte-identical;
- partition groups do not overlap;
- normalizers use training only;
- example target timing matches force-time oracle labels.

### Shape/mask contract

- variable histories and candidates;
- H1 contains no illegal transition features;
- padding masks prevent padded agents/candidates from affecting outputs;
- empty/unknown-only candidate sets are valid;
- outputs normalize and covariance is positive definite.

### Residual identity

Zero-initialized residual head reproduces the rule baseline exactly within tolerance; this proves composition semantics.

### Checkpoint compatibility

- save/load round trip;
- wrong feature/schema/normalizer/model hashes fail closed;
- no checkpoint + evidence mode fails rather than using untrained smoke;
- diagnostic untrained smoke is clearly labelled and cannot enter confirmation reports.

### Training determinism

Two bounded CPU runs with fixed seeds produce identical or documented tolerance-equivalent losses/checkpoint outputs. If device kernels prevent bit identity, state and test numerical tolerance and deterministic configuration.

### Calibration

Golden logits/covariances verify temperature/covariance scaling and no final-test fitting.

### Censoring and unavailable labels

Losses and metrics use correct masks/denominators; no NaN or silent row dropping.

### Runtime fallback

Missing/low-confidence belief uses the explicit rule fallback and records provenance. Fallback outputs are not counted as learned-model activation.

### Online adaptation safety

Updates remain bounded, use no oracle labels, rollback correctly, and leave planner actions unchanged in shadow mode.

## Training and verification commands

```bash
uv run pytest \
  tests/prediction/test_goal_belief_dataset.py \
  tests/prediction/test_learned_goal_belief.py \
  tests/prediction/test_learned_goal_belief_trainer.py \
  tests/prediction/test_goal_belief_checkpoint_contract.py \
  tests/prediction/test_goal_belief_calibration.py \
  tests/prediction/test_goal_belief_online_adapter.py -q

uv run ruff check robot_sf/prediction scripts/data scripts/training tests/prediction
uv run ruff format --check <changed paths>
git diff --check
```

Required bounded workflow:

```bash
uv run python scripts/data/build_goal_belief_dataset.py --config <smoke-config>
uv run python scripts/training/train_goal_belief.py --config <cpu-smoke-config>
uv run python scripts/benchmark/run_goal_force_inference.py --config <frozen-learned-confirmation-config>
BASE_REF=origin/main scripts/dev/pr_ready_check.sh
```

Full training must use a registered launch/config/manifest and produce a reproducible evidence artifact; do not place ad-hoc checkpoint paths into benchmark configs.

## Acceptance criteria

- [ ] Actor features and oracle labels are physically separated and leakage-tested.
- [ ] Dataset partitions are episode/map/scenario aware and manifest/checksum bound.
- [ ] Model is explicitly residual to the frozen rule baseline.
- [ ] H1–H16 causal histories and variable candidate masks are supported.
- [ ] Losses handle censoring, unavailable labels, calibration, and class imbalance truthfully.
- [ ] Checkpoint/normalization/schema compatibility fails closed.
- [ ] Learned model is compared against the exact best rule baseline and required ablations.
- [ ] Final test configuration and calibration are frozen before execution.
- [ ] Proper scores/calibration, change metrics, unknown detection, runtime, and robustness satisfy preregistered gates or the result is reported as non-promotable.
- [ ] Online adaptation remains shadow-only and bounded.
- [ ] Focused/regression tests and PR-ready checks pass.

## Stop rule

Do not promote the learned backend if it improves only point error while worsening proper scores/calibration, relies on oracle leakage, cannot beat the frozen rule baseline on held-out groups, exceeds runtime, or increases unsafe fallback ambiguity. Retain the rule model and publish the negative diagnostic result.

## Evidence and claim boundary

This issue may establish simulator-domain learned-model performance against held-out Robot SF traces. It does not establish human-intention validity, real-world transfer, planner/RL improvement, or paper/dissertation evidence without the final frozen-predictor campaign and a separate claim ruling.

```yaml
schema: goal_autopilot_preparation.v1
repository: ll7/robot_sf_ll7
issue: 8098
source_body_sha256: 41aba641d67f8ab96d4667f3cb04d97f7e99cf53913c19677bd97f8865f1d3c0
source_comments_sha256: 
audit_schema: open_issue_contract_audit.v1
audit_digest: 3f3c51addec2c0ac8aad5ca07f84a63669ce53bf2603a0919267fd1bd198b7c6
audit_classification: state_conflict
next_action: reconcile_state_labels
authority: lifecycle_owner
execution_mode: lifecycle
preferred_worker: MaxRunner
expected_pr_runner_label: runner:max
implementation_admitted: False
state_ready_change_proposed: False
mutation_batch: open-issues-final-20260831
```

This packet is preparation evidence only. It never overrides live labels, exact claim state, branch state, typed dependencies, domain gates, compute authority, release authority, or scientific evidence rules.
